src/Init/Data/Char/Basic.lean

```diff
@@ -17,17 +18,19 @@ instance : HasSizeof Char :=
 ⟨fun c => c.val.toNat⟩
 
 namespace Char
-def utf8Size (c : Char) : UInt32 :=
-let v := c.val;
-     if UInt32.land v 0x80 = 0    then 1
-else if UInt32.land v 0xE0 = 0xC0 then 2
-else if UInt32.land v 0xF0 = 0xE0 then 3
-else if UInt32.land v 0xF8 = 0xF0 then 4
-else if UInt32.land v 0xFC = 0xF8 then 5
-else if UInt32.land v 0xFE = 0xFC then 6
-else if v = 0xFF                  then 1
+def utf8Size' (v : UInt8) : UInt32 :=
+     if UInt8.land v 0x80 = 0    then 1
+else if UInt8.land v 0xE0 = 0xC0 then 2
+else if UInt8.land v 0xF0 = 0xE0 then 3
+else if UInt8.land v 0xF8 = 0xF0 then 4
+else if UInt8.land v 0xFC = 0xF8 then 5
+else if UInt8.land v 0xFE = 0xFC then 6
+else if v = 0xFF                 then 1
 else 0
 
+def utf8Size (c : Char) : UInt32 :=
+utf8Size' c.val.toUInt64.toUInt8
```

Comment:
I think you've actually uncovered a bug here: getting the UTF8 encoding size from a codepoint (`utf8Size`), and from the first byte of its encoding (`utf8Size'`) are not the same operation. The former should use column # 3 of https://en.wikipedia.org/wiki/UTF-8#Description, while the latter is # 5 (as implemented). It should not be possible to implement one in terms of the other.
@leodemoura agree?

Reply:
@Kha I agree. That being said, I am confused by the current `utf8Size` code. The table on wikipedia says the size is between 1 and 4, but the function also returns 5 and 6. Just to make sure we are all on the same page. The implementation should be:
```lean
def utf8Size (c : Char) : UInt32 :=
let v := c.val;
if v < 0x7F then 1
else if v < 0x7FF then 2
else if v < 0xFFFF then 3
else 4
```

Reply:
@Kha do you agree?

Reply:
Agreed. As the Wiki page states, Unicode was restricted to 21 bits, i.e. 4 UTF8 bytes, in 2003.